istioctl/pkg/tag/util.go

```diff
@@ -0,0 +1,171 @@
+	})
+	if err != nil {
+		return nil, err
+	}
+	return webhooks.Items, nil
+}
+
+// GetNamespacesWithTag retrieves all namespaces pointed at the given tag.
+func GetNamespacesWithTag(ctx context.Context, client kubernetes.Interface, tag string) ([]string, error) {
+	namespaces, err := client.CoreV1().Namespaces().List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", label.IoIstioRev.Name, tag),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	nsNames := make([]string, len(namespaces.Items))
+	for i, ns := range namespaces.Items {
+		nsNames[i] = ns.Name
+	}
+	return nsNames, nil
+}
+
+// GetWebhookName extracts tag name from webhook object.
+func GetWebhookName(wh admit_v1.MutatingWebhookConfiguration) (string, error) {
+	if tagName, ok := wh.ObjectMeta.Labels[IstioTagLabel]; ok {
+		return tagName, nil
+	}
+	return "", fmt.Errorf("could not extract tag name from webhook")
+}
+
+// GetWebhookRevision extracts tag target revision from webhook object.
+func GetWebhookRevision(wh admit_v1.MutatingWebhookConfiguration) (string, error) {
+	if tagName, ok := wh.ObjectMeta.Labels[label.IoIstioRev.Name]; ok {
+		return tagName, nil
+	}
+	return "", fmt.Errorf("could not extract tag revision from webhook")
+}
+
+// DeleteTagWebhooks deletes the given webhooks.
+func DeleteTagWebhooks(ctx context.Context, client kubernetes.Interface, webhooks []admit_v1.MutatingWebhookConfiguration) error {
+	var result error
+	for _, wh := range webhooks {
+		result = multierror.Append(client.AdmissionregistrationV1().MutatingWebhookConfigurations().Delete(ctx, wh.Name, metav1.DeleteOptions{})).ErrorOrNil()
+	}
+	return result
+}
+
+var neverMatch = &metav1.LabelSelector{
+	MatchLabels: map[string]string{
+		"istio.io/deactivated": "never-match",
+	},
+}
+
+// DefaultRevisionExists checks whether there is an existing default revision, either implicit by virtue of having a
+// previous non-revisioned installation or an explicit default tag. Should be used in installer when deciding whether
+// to make an installation the default.
+func DefaultRevisionExists(ctx context.Context, client kubernetes.Interface) (bool, error) {
+	tagWhs, err := GetWebhooksWithTag(ctx, client, DefaultRevisionName)
+	if err != nil {
+		return false, err
+	}
+	whs, err := GetWebhooksWithRevision(ctx, client, DefaultRevisionName)
+	if err != nil {
+		return false, err
+	}
+	return len(tagWhs) > 0 || len(whs) > 0, nil
+}
+
+// DeactivateIstioInjectionWebhook deactivates the istio-injection webhook from the given MutatingWebhookConfiguration if exists.
+// used rather than just deleting the webhook since we want to keep it around after changing the default so user can later
+// switch back to it. This is a hack but it is meant to cover a corner case where a user wants to migrate from a non-revisioned
+// old version and then later decides to switch back to the old revision again.
+func DeactivateIstioInjectionWebhook(ctx context.Context, client kubernetes.Interface) error {
```

Comment:
This tries to make it possible to switch back to a non-revisioned install after making a revisioned install the default, a bit gross

Reply:
Right, but I can't think of a better option. For 1.11+ we should just not install this webhook, even for the 'revision=""' - so it will not be a long-term problem, once 1.10 is out of support we can remove this hack.

Make sure you document the behavior - and the fact that if user make changes to the 1.10 'unrevisioned', like enable access logs - it will revert your hack. 

It may be a case of just documenting a manual rollback procedure for 1.10 - and making sure in future we are fine.